File: jas/canvas/pane_rendering_test.py
"""Tests for pane rendering helpers."""

from workspace.pane import (
    PaneLayout, PaneKind, EdgeSide, SnapConstraint, WindowTarget,
    DEFAULT_TOOLBAR_WIDTH,
)
from canvas.pane_rendering import (
    compute_pane_geometries, compute_shared_borders, compute_snap_lines,
)


def test_geometries_from_default_layout():
    pl = PaneLayout.default_three_pane(1000, 700)
    geos = compute_pane_geometries(pl)
    assert len(geos) == 3
    assert all(g.visible for g in geos)


def test_geometries_pane_positions():
    pl = PaneLayout.default_three_pane(1000, 700)
    geos = compute_pane_geometries(pl)
    toolbar = next(g for g in geos if g.kind == PaneKind.TOOLBAR)
    canvas = next(g for g in geos if g.kind == PaneKind.CANVAS)
    dock = next(g for g in geos if g.kind == PaneKind.DOCK)
    assert toolbar.x == 0
    assert toolbar.width == DEFAULT_TOOLBAR_WIDTH
    assert abs(canvas.x - (toolbar.x + toolbar.width)) < 0.001
    assert abs(dock.x - (canvas.x + canvas.width)) < 0.001
    assert toolbar.height == 700
    assert canvas.height == 700
    assert dock.height == 700


def test_geometries_canvas_maximized():
    pl = PaneLayout.default_three_pane(1000, 700)
    pl.toggle_canvas_maximized()
    geos = compute_pane_geometries(pl)
    canvas = next(g for g in geos if g.kind == PaneKind.CANVAS)
    assert canvas.x == 0 and canvas.y == 0
    assert canvas.width == 1000 and canvas.height == 700
    assert canvas.z_index == 0


def test_geometries_hidden_pane():
    pl = PaneLayout.default_three_pane(1000, 700)
    pl.hide_pane(PaneKind.TOOLBAR)
    geos = compute_pane_geometries(pl)
    toolbar = next(g for g in geos if g.kind == PaneKind.TOOLBAR)
    assert not toolbar.visible
    canvas = next(g for g in geos if g.kind == PaneKind.CANVAS)
    assert canvas.visible


def test_geometries_z_order():
    pl = PaneLayout.default_three_pane(1000, 700)
    geos = compute_pane_geometries(pl)
    canvas = next(g for g in geos if g.kind == PaneKind.CANVAS)
    toolbar = next(g for g in geos if g.kind == PaneKind.TOOLBAR)
    dock = next(g for g in geos if g.kind == PaneKind.DOCK)
    assert canvas.z_index < toolbar.z_index < dock.z_index


def test_shared_borders_default():
    pl = PaneLayout.default_three_pane(1000, 700)
    borders = compute_shared_borders(pl)
    # toolbar-canvas and canvas-dock borders
    assert len(borders) == 2
    for b in borders:
        assert b.is_vertical
        assert b.height == 700


def test_no_borders_when_maximized():
    pl = PaneLayout.default_three_pane(1000, 700)
    pl.toggle_canvas_maximized()
    assert compute_shared_borders(pl) == []


def test_snap_lines_computation():
    pl = PaneLayout.default_three_pane(1000, 700)
    cid = pl.pane_by_kind(PaneKind.CANVAS).id
    preview = [
        SnapConstraint(cid, EdgeSide.LEFT, WindowTarget(EdgeSide.LEFT)),
        SnapConstraint(cid, EdgeSide.TOP, WindowTarget(EdgeSide.TOP)),
    ]
    lines = compute_snap_lines(preview, pl)
    assert len(lines) == 2
    assert any(l.width == 4 and l.height > 4 for l in lines)


def test_geometries_from_none():
    assert compute_pane_geometries(None) == []

File: jas_ocaml/test/canvas/pane_rendering_test.ml
open Jas.Pane
open Jas.Pane_rendering

let near a b = abs_float (a -. b) < 0.001

let tests = [
  Alcotest.test_case "geometries_from_default_layout" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    let geos = pane_geometries pl in
    assert (List.length geos = 3);
    assert (List.for_all (fun g -> g.visible) geos));

  Alcotest.test_case "geometries_pane_positions" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    let geos = pane_geometries pl in
    let toolbar = List.find (fun g -> g.kind = Toolbar) geos in
    let canvas = List.find (fun g -> g.kind = Canvas) geos in
    let dock = List.find (fun g -> g.kind = Dock) geos in
    assert (toolbar.x = 0.0);
    assert (toolbar.width = default_toolbar_width);
    assert (near canvas.x (toolbar.x +. toolbar.width));
    assert (near dock.x (canvas.x +. canvas.width));
    assert (toolbar.height = 700.0);
    assert (canvas.height = 700.0);
    assert (dock.height = 700.0));

  Alcotest.test_case "geometries_canvas_maximized" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    toggle_canvas_maximized pl;
    let geos = pane_geometries pl in
    let canvas = List.find (fun g -> g.kind = Canvas) geos in
    assert (canvas.x = 0.0);
    assert (canvas.y = 0.0);
    assert (canvas.width = 1000.0);
    assert (canvas.height = 700.0);
    assert (canvas.z_index = 0));

  Alcotest.test_case "geometries_hidden_pane" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    hide_pane pl Toolbar;
    let geos = pane_geometries pl in
    let toolbar = List.find (fun g -> g.kind = Toolbar) geos in
    assert (not toolbar.visible);
    let canvas = List.find (fun g -> g.kind = Canvas) geos in
    assert canvas.visible);

  Alcotest.test_case "geometries_z_order" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    let geos = pane_geometries pl in
    let canvas = List.find (fun g -> g.kind = Canvas) geos in
    let toolbar = List.find (fun g -> g.kind = Toolbar) geos in
    let dock = List.find (fun g -> g.kind = Dock) geos in
    assert (canvas.z_index < toolbar.z_index);
    assert (toolbar.z_index < dock.z_index));

  Alcotest.test_case "shared_borders_default" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    let borders = shared_borders pl in
    (* toolbar|canvas and canvas|dock borders *)
    assert (List.length borders = 2);
    List.iter (fun b -> assert b.is_vertical; assert (b.bh = 700.0)) borders);

  Alcotest.test_case "no_borders_when_maximized" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    toggle_canvas_maximized pl;
    let borders = shared_borders pl in
    assert (borders = []));

  Alcotest.test_case "snap_lines_computation" `Quick (fun () ->
    let pl = default_three_pane ~viewport_w:1000.0 ~viewport_h:700.0 in
    let canvas_id = (Option.get (pane_by_kind pl Canvas)).id in
    let preview = [
      { snap_pane = canvas_id; edge = Left; target = Window_target Left };
      { snap_pane = canvas_id; edge = Top; target = Window_target Top };
    ] in
    let lines = snap_lines preview pl in
    assert (List.length lines = 2);
    (* Left edge snap line should be vertical (narrow width, tall height) *)
    assert (List.exists (fun l -> l.lw = 4.0 && l.lh > 4.0) lines));
]

let () =
  Alcotest.run "PaneRendering" [
    "Pane rendering tests", tests;
  ]
